### PR TITLE
scatter chart for RE

### PR DESCRIPTION
**Pull request by ivanbasov**

### CODE REVIEW

**Review by msoeken — APPROVED** [review]

**Comment by ivanbasov** [comment]
@billti and @swernli , could you please take a look?

I have recently added Jupyter notebook support and removed a reference to an external colormap package. A Jupyter notebook sample is also there.

Besides careful testing, the only item to add seems to be unit tests for ticks.tsx. Hope to proceed with them soon.

File: widgets/js/index.tsx
Comment (billti) [inline widgets/js/index.tsx:47]: Remove these console.error logs.

File: npm/ux/colormap.tsx
Comment (billti) [inline npm/ux/colormap.tsx:22]: How does anyone supply their own map?

File: npm/ux/data.tsx
Comment (billti) [inline npm/ux/data.tsx:3]: Add a newline between the license header and any other comments.

File: npm/ux/generate_report_code.py
Comment (billti) [inline npm/ux/generate_report_code.py:1]: I'm not clear what uses this file? Does any other code use it? If it's something that needs to be run manually, how? It could do with some notes/instructions.

Comment (billti) [reply npm/ux/generate_report_code.py:1]: (And a license header)

File: widgets/js/index.tsx
Comment (billti) [inline widgets/js/index.tsx:94]: Ideally the render call in the widget should just render the component - similar to those above. Does this next 20 lines of logic belong here or in the component? Do we need a new component to wrap it up this logic and the two child components maybe? (Just thinking here - not proposing either of these yet).

File: npm/ux/index.ts
Comment (billti) [inline npm/ux/index.ts:15]: Do all of these need to be exported at the top level of the package? Seems like some of these should be internal to the components. (e.g. does a consumer need to worry about xAxis and yAxis?)

File: npm/ux/colormap.tsx
Comment (minestarks) [inline npm/ux/colormap.tsx:5]: Don't we need license notices if we're copying code?

File: widgets/js/index.tsx
Comment (minestarks) [inline widgets/js/index.tsx:105]: Wouldn't this start to look a bit weird after `idx` exceeds 26? Is it important to use letters or can we use a number suffix?

Comment (ivanbasov) [reply widgets/js/index.tsx:105]: Good question! Thank you, @minestarks ! On one hand, the chart and the table could show many dozens of series and rows correspondingly. We even prepare colormap for the purpose. However, it is pretty hard to analyze the chart and the summary table with the number of series/rows greater than 10-15, ok 20. So, technically we could replace (a), (b), (c) with (1), (2), (3). But do we want to support such scenarios? Or maybe we could show (a), (b), (c) if the number of series/rows is less than 27, and numbers otherwise.

We have the same code on VS Code extension side (this one is a copy for Jupyter widgets).
@billti and @minestarks , what do you think?

File: npm/ux/colormap.tsx
Comment (ivanbasov) [reply npm/ux/colormap.tsx:45]: Do we need to support custom maps?
1. Map definitions can be provided as arrays for colors with 'distances' between them to set interpolations. It seems to be a complication for the API.
2. Supporting custom maps also requires some kind of error handling for situations such as the number of map intervals is greater than the number of colors need to produce.
3. Most of maps provided in colormap original samples provide shades of few number of gradients. It seems to be not applicable for the scatter chart purposes.

I do not think that providing custom color maps would be beneficial for customers.
Maybe we could supply custom colors for individual series at the Jupyter notebook scenario? E.g. draw majorana with '#00FF00' and gate-based with '#FF0000'.

What do you think @billti?

Comment (ivanbasov) [reply npm/ux/colormap.tsx:45]: Added support of custom colors for the Jupyter notebook scenario. Hopefully, this provides enough flexibility for users.

File: widgets/js/index.tsx
Comment (ivanbasov) [reply widgets/js/index.tsx]: Here are my thoughts:
1. for VS Code scenario labels like "BitFlip (a)", "BitFlip (b)" are enough because we would consider a few items in the run (now just 8) and add "-2", "-3" for next runs.
2. For Jupyter notebook scenario, it is worth to have numbers (not letters for easiest correspondence with params sent for the estimation). I have added this starting with 1. Please let me know if you think we should start with 0.
3. For Jupyter scenario, it could be helpful to show custom labels for runNames. I have added this as well and provided in a sample.

File: npm/ux/report.tsx
Comment (ivanbasov) [inline npm/ux/report.tsx:239]: still debugging here

File: resource_estimator/src/estimates/data/job_params.rs
Comment (ivanbasov) [inline resource_estimator/src/estimates/data/job_params.rs:193]: @msoeken , do you agree it is a good time to make it the default option?

File: pip/tests/test_re.py
Comment (swernli) [inline pip/tests/test_re.py:103]: Hmm... I think the fact that this test had to change is a strong indication that we don't want to change the default in Python. This pattern here is used in notebook samples and docs pages, which means this would be a **breaking change** for Python consumers.

Comment (ivanbasov) [reply pip/tests/test_re.py:103]: Sure. This is a breaking change for code we just released. It is not great to make the change the next day. But if we ever want to switch it could be the only option. So, let us wait for more news today to make the final decision: if we want the frontier by default anywhere either in Python or in VS Code.

**Comment by swernli** [comment]
I think a merge from main might help with the memory task, as main now skips writing the comment if the PR doesn't change memory values and this PR does not change them.

**Comment by swernli** [comment]
Oh! This PR is coming from a fork. I think that is affecting how the task is able to write back. Take a look at [the logs for the benchmark task](https://github.com/microsoft/qsharp/actions/runs/7547644236/job/20559372973#step:3:454) that show the same failure, but that task treats it as a warning instead of a failure:

> Warning: Failed to comment: HttpError: Resource not accessible by integration
Commenting is not possible from forks.

@sezna I think you'll need to adjust the behavior of the memory task as well, otherwise it will continue to fail when PRs come from forks.

@ivanbasov you should be able to get into a good state by instead creating your branch in this repo rather than in your fork and then opening the PR from there. Then you can close this PR.

File: resource_estimator/src/estimates/data/job_params.rs
Comment (ivanbasov) [reply resource_estimator/src/estimates/data/job_params.rs:193]: Decided to return the API back to SinglePoint by default but use the Frontier as the only option in the VS Code scenaition

File: pip/tests/test_re.py
Comment (ivanbasov) [reply pip/tests/test_re.py]: agreed on defaults. reverted to agreed ones

File: samples/estimation/estimation-frontier-widgets.ipynb
Comment (billti) [inline samples/estimation/estimation-frontier-widgets.ipynb:16]: Do you do anything with result1? I don't see it referenced anywhere else. It seems to be dead code/unused and this line should be removed.

File: npm/src/ux/utils.ts
Comment (billti) [inline npm/src/ux/utils.ts:4]: Not sure we need either of these methods. The number formatter with options such as 'maximumFractionDigits', and includes locale-specific separators by default, has been widely implemented for about 10 years across browsers.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat

File: npm/test/basics.js
Comment (billti) [inline npm/test/basics.js:898]: Am I reading the above tests right that 
 - if my data runs 1 to 9, I just get a tick at 1
 - if my data runs 2 to 10, I just get a tick at 10
 - if my data runs 2 to 9, I get 8 ticks from 2 through 9

Maybe that 'feels' right in practice, but seems a little odd.

File: npm/test/basics.js
Comment (billti) [inline npm/test/basics.js:928]: Should we prefix the singular with the value as we do for the plural (e.g. "1 nanosecond")?

File: npm/test/basics.js
Comment (billti) [inline npm/test/basics.js:1104]: For readability, can we have all these consts above all the code where they're used, rather than after.

File: npm/test/basics.js
Comment (ivanbasov) [reply npm/test/basics.js:898]: Thank you, @billti! The intention was to show 1, 10, 100, ... if possible. If at least one of them is shown, then do not show anything else. Otherwise, show intermediate results on the lower level, e.g. 2, 3, 4, 5; or 50, 60, 70, 80, 90; or 34, 35, 36 etc.
Yes, in the "worst" case it could be 2 through 9 or similar 8 numbers. Could be worse in case of 1, 10, ... 1'000'000'000, etc.

So, the number of ticks could be from 1 to 8. My original goal was to find at least a single tick. Everything else was secondary. Maybe we could set more elegant goals like to find from 2 to 5 ticks or so. Thoughts?

File: npm/ux/data.ts
Comment (billti) [inline npm/ux/data.ts:27]: If this doesn't need to be provided, then just mark it optional, i.e. parameter would be declared `frontierEntryIndex?: number` (see https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters)

File: npm/ux/data.ts
Comment (billti) [inline npm/ux/data.ts:53]: So here it's returning one frontier entry, but you still set a property called 'frontierEntries' which is all the frontierEntries. Is that needed and used for anything?

File: npm/ux/data.ts
Comment (billti) [inline npm/ux/data.ts:25]: Should this just be called `GetFrontierEntry` and require the second param? There's little point in calling it unless you pass the second param to get the specific entry, as it just reflects the first param directly.

File: vscode/src/webview/rePage.tsx
Comment (billti) [inline vscode/src/webview/rePage.tsx:63]: We might want a different name than `Summary` for the custom component, as `summary` is a standard HTML tag that is also used in this code (and is something very different). I was confused for a second until I realized that distinction.

File: npm/test/basics.js
Comment (ivanbasov) [reply npm/test/basics.js:938]: I thought to follow https://quantum.microsoft.com/en-us/experience/quantum-cryptography where we use units without values. But I do not have a strong preference.

File: npm/package.json
Comment (ivanbasov) [reply npm/package.json]: Tried once again. Unfortunately, basics.js could not refer to a typescript file. So, likely this solution with moving ticks.ts to ./src/ux (discussed a week ago with @billti  and @minestarks ) is the only solution to provide unit tests for this ts file. Please let me know if there are more ideas to try.

File: npm/ux/data.ts
Comment (ivanbasov) [reply npm/ux/data.ts]: Actually, ReData was used for two different purposes: 
- original binding strucutre for controls
- structure received from the RE, which was extended with FrontierEntries

Now, I have intorduced SingleEstimateResult for the former purpose. Hopefully, it would clarify data conversions.

File: npm/test/basics.js
Comment (ivanbasov) [reply npm/test/basics.js:898]: Let us consider this aside from the current PR. I have created https://github.com/microsoft/qsharp/issues/1036

File: npm/package.json
Comment (billti) [reply npm/package.json]: OK. I see what you were trying to do now. If it's working, let's get this in now and clean it up after. It is a bit messy/confusing when you have compiled JS, source only TSX, and package exports mixed together.

File: npm/ux/scatterChart.tsx
Comment (billti) [inline npm/ux/scatterChart.tsx:8]: I think this should just be a relative path to the source, then you can remove the package.json changes to exports entirely, i.e.

`} from "../src/ux/ticks.js"`

(Note: Still use a .js extension and it will resolve).

This way, the unit tests in basics.js are still testing the compiled source under 'dist', but any package importing the components as source are still only pulling source, and aren't pulling a mix of source tsx and built js for the ux components.

Not sure if that makes sense, but try it. (Or leave as-is, and I'll do a follow up)

Comment (ivanbasov) [reply npm/ux/scatterChart.tsx:8]: Let me try

**Review by billti — APPROVED** [review]

File: npm/ux/scatterChart.tsx
Comment (ivanbasov) [reply npm/ux/scatterChart.tsx]: Seems to work. Thank you, @billti !

**Review by DmitryVasilevsky — APPROVED** [review]

**Pull request merged**
